### PR TITLE
Validating against decrementing LatestVersion of deployment config

### PR DESCRIPTION
**Pull request by abhgupta**
Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1230675

This is a validation to prevent users from decrementing the LatestVersion of a deployment config as it causes multiple issues. While the bug covers one of the issues, other issues that would manifest would be failures to trigger new deployments and rollback since these features rely on a new deployment being kicked off by simply incrementing the LatestVersion of the deployment config by 1.

Validations added:
 - LatestVersion cannot be negative
 - LatestVersion cannot be decremented
 - LatestVersion can be incremented by only 1

### CODE REVIEW

**Comment by abhgupta** [comment]
@ironcladlou PTAL

File: pkg/deploy/api/validation/validation.go [hunk 1]
```diff
@@ -32,6 +32,9 @@ func ValidateDeploymentConfigUpdate(newConfig *deployapi.DeploymentConfig, oldCo
 	allErrs := fielderrors.ValidationErrorList{}
 	allErrs = append(allErrs, validation.ValidateObjectMetaUpdate(&newConfig.ObjectMeta, &oldConfig.ObjectMeta).Prefix("metadata")...)
 	allErrs = append(allErrs, ValidateDeploymentConfig(newConfig)...)
+	if newConfig.LatestVersion < oldConfig.LatestVersion {
```
Comment (liggitt) [inline pkg/deploy/api/validation/validation.go]: should `ValidateDeploymentConfig` make sure this is never `< 0`? would that cause any issues?

Comment (liggitt) [inline pkg/deploy/api/validation/validation.go]: also, should it ever increment more than 1?

Comment (ironcladlou) [inline pkg/deploy/api/validation/validation.go]: It should always be >= 0, and although gaps resulting from increments of >1 shouldn't really be a problem functionally, I think it makes sense to keep it as contiguous as possible.

Comment (liggitt) [inline pkg/deploy/api/validation/validation.go]: just thinking of edge cases like range exhaustion... first deployment: 1, second deployment: 2147483648, third deployment: oops

Comment (liggitt) [inline pkg/deploy/api/validation/validation.go]: I can't think of a case where a well-behaved updater is going to increment more than 1 over the currently stored value...

Comment (smarterclayton) [inline pkg/deploy/api/validation/validation.go]: Upstream is adding sequence number which will eventually supersede this.  But I don't know of any reasonable case where we should allow anything other than a monotonic increment +1.  You can potentially ignore greater increments during update.

TBH I don't think this is under the user's control.  Every update of the spec should be +1.  Also, the initial creation of the deployment config should probably be 1, not 0.  But there is a lot of code here to change.  Original design is that increments of latest version are the things that result in deployments.  Changes that shouldn't result in a deployment don't increment version.  Decision about whether to increment version is not under client control.

**Comment by abhgupta** [comment]
For now, I can:

1. Block negative values for LatestVersion
2. Enforce increments by 1 (if an update has an increment of 2+, I can implicitly dial it back to an increment of just 1)

We can consider taking away this attribute from user control once the sequence number functionality lands in upstream.

Do we have agreement?

**Comment by abhgupta** [comment]
If the user specified LatestVersion is an increment of 2+, can I override it in the validation? If not, I am really not in favor or bringing in admission control for this and would rather just block it and return an error back to the user.

**Comment by liggitt** [comment]
> 1. Block negative values for LatestVersion

Yes

> 2. Enforce increments by 1

I think you should fail validation... @ironcladlou any valid use case to the contrary you can think of?

**Comment by abhgupta** [comment]
@liggitt Sure, I like that better as it saves me from implicitly changing user provided values insdie validation code.

**Comment by abhgupta** [comment]
Made the changes and specified the validations that we are now performing in the description of this PR.

**Comment by abhgupta** [comment]
@smarterclayton Please review/merge

**Comment by abhgupta** [comment]
[test]

File: pkg/deploy/api/validation/validation.go [hunk 2]
```diff
@@ -32,6 +35,11 @@ func ValidateDeploymentConfigUpdate(newConfig *deployapi.DeploymentConfig, oldCo
 	allErrs := fielderrors.ValidationErrorList{}
```
Comment (ironcladlou) [inline pkg/deploy/api/validation/validation.go]: I think there's a gap in my validation/persistence framework knowledge here... is `ValidateDeploymentConfigUpdate` called reflectively by the generic REST stuff?

Comment (liggitt) [inline pkg/deploy/api/validation/validation.go]: Called by way of the update strategy

**Comment by abhgupta** [comment]
@danmcp @smarterclayton I would like to make a case for letting this one in. This should be a simple validation and will help avoid confusion in cases of both triggered and manual deployments (they might not happen) if the user decrements the LatestVersion.

**Comment by smarterclayton** [comment]
Approved, but want @liggitt's sign off as well.

**Comment by liggitt** [comment]
LGTM

**Comment by smarterclayton** [comment]
[merge]

**Pull request merged**